appinventor/appengine/src/com/google/appinventor/client/TopToolbar.java

```diff
@@ -208,95 +126,31 @@ public void updateMenuState(int numSelectedProjects, int numProjects) {
-          new UploadKeystoreAction()));
-    }
-    fileItems.add(new DropDownItem(WIDGET_NAME_DOWNLOAD_KEYSTORE, MESSAGES.downloadKeystoreMenuItem(),
-        new DownloadKeystoreAction()));
-    if (!isReadOnly) {
-      fileItems.add(new DropDownItem(WIDGET_NAME_DELETE_KEYSTORE, MESSAGES.deleteKeystoreMenuItem(),
-          new DeleteKeystoreAction()));
-    }
-    refreshMenu(fileDropDown, fileItems);
-  }
+  public TopToolbar() {
+    // Should the UI be in read only mode?
+    isReadOnly = Ode.getInstance().isReadOnly();
 
-  private void createConnectMenu() {
-    List<DropDownItem> connectItems = Lists.newArrayList();
-    connectItems.add(new DropDownItem(WIDGET_NAME_WIRELESS_BUTTON,
-        MESSAGES.AICompanionMenuItem(), new WirelessAction()));
-    if (iamChromebook) {
-      connectItems.add(new DropDownItem(WIDGET_NAME_CHROMEBOOK,
-          MESSAGES.chromebookMenuItem(), new ChromebookAction()));
-    } else {
-      connectItems.add(new DropDownItem(WIDGET_NAME_EMULATOR_BUTTON,
-          MESSAGES.emulatorMenuItem(), new EmulatorAction()));
-      connectItems.add(new DropDownItem(WIDGET_NAME_USB_BUTTON, MESSAGES.usbMenuItem(),
-          new UsbAction()));
-    }
-    connectItems.add(null);
-    connectItems.add(new DropDownItem(WIDGET_NAME_REFRESHCOMPANION_BUTTON, MESSAGES.refreshCompanionMenuItem(),
-            new RefreshCompanionAction()));
-    connectItems.add(null);
-    connectItems.add(new DropDownItem(WIDGET_NAME_RESET_BUTTON, MESSAGES.resetConnectionsMenuItem(),
-        new ResetAction()));
-    if (!iamChromebook) {
-      connectItems.add(new DropDownItem(WIDGET_NAME_HARDRESET_BUTTON, MESSAGES.hardResetConnectionsMenuItem(),
-          new HardResetAction()));
-    }
-    refreshMenu(connectDropDown, connectItems);
-  }
+    initWidget(UI_BINDER.createAndBindUi(this));
 
-  private void createBuildMenu() {
-    List<DropDownItem> buildItems = Lists.newArrayList();
-    buildItems.add(new DropDownItem(WIDGET_NAME_BUILD_ANDROID_APK, MESSAGES.showExportAndroidApk(),
-        new BarcodeAction(false, false)));
-    buildItems.add(new DropDownItem(WIDGET_NAME_BUILD_ANDROID_AAB, MESSAGES.showExportAndroidAab(),
-        new BarcodeAction(false, true)));
+    if (isReadOnly) {
+      fileDropDown.removeItemById(WIDGET_NAME_NEW);
+      fileDropDown.removeItemById(WIDGET_NAME_IMPORTPROJECT);
+      fileDropDown.removeItemById(WIDGET_NAME_IMPORTTEMPLATE);
+      fileDropDown.removeItemById(WIDGET_NAME_DELETE);
+      fileDropDown.removeItemById(WIDGET_NAME_SAVE);
+      fileDropDown.removeItemById(WIDGET_NAME_SAVE_AS);
+      fileDropDown.removeItemById(WIDGET_NAME_CHECKPOINT);
+      fileDropDown.removeItemById(WIDGET_NAME_UPLOAD_KEYSTORE);
+      fileDropDown.removeItemById(WIDGET_NAME_DELETE_KEYSTORE);
+      // TODO: Are these implemented?
+      fileDropDown.removeItemById(WIDGET_NAME_BUILD_ANDROID_APK);
+      fileDropDown.removeItemById(WIDGET_NAME_BUILD_ANDROID_AAB);
```

Comment:
The APK/AAB options will be in the build menu, not the file menu

Reply:
> The APK/AAB options will be in the build menu, not the file menu

This turns out to be a production bug. On ai2, those options are present on the build menu, it's possible to build apps in readOnly mode. I changed it in my most recent commit, but that empties out the entire build menu. If we really don't think it should be possible to build in readOnly mode, I should probably remove the entire drop-down.